pkg/gadget-service/instance-manager/manager.go

```diff
@@ -0,0 +1,201 @@
+
+	runtime runtime.Runtime
+	store   Store
+
+	Service
+}
+
+func New(runtime runtime.Runtime, options ...Option) (*Manager, error) {
+	mgr := &Manager{
+		gadgetInstances: make(map[string]*GadgetInstance),
+		gadgetDone:      make(chan bool, 1),
+		runtime:         runtime,
+	}
+	for _, opt := range options {
+		err := opt(mgr)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if mgr.store == nil {
+		return nil, errors.New("no store provided")
+	}
+	return mgr, nil
+}
+
+func (p *Manager) Store() Store {
+	return p.store
+}
+
+// StopGadget cancels a running gadget, but leaves the results accessible
+func (p *Manager) StopGadget(id string) error {
+	log.Printf("stopping gadget %q", id)
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	gadgetInstance, ok := p.gadgetInstances[id]
+	if !ok {
+		return fmt.Errorf("gadget not found")
+	}
+	gadgetInstance.cancel()
+
+	// ensure that the gadget is stopped before returning
+	if gadgetInstance.state == stateRunning {
+		<-p.gadgetDone
+	}
+
+	return nil
+}
+
+// RemoveGadget cancels and removes a gadget
+func (p *Manager) RemoveGadget(id string) error {
+	log.Printf("removing gadget %q", id)
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	gadgetInstance, ok := p.gadgetInstances[id]
+	if !ok {
+		return fmt.Errorf("gadget not found")
+	}
+	gadgetInstance.cancel()
+	delete(p.gadgetInstances, id)
+	return nil
+}
+
+func (p *Manager) RunGadget(
+	instance *api.GadgetInstance,
+) {
+	ctx, cancel := context.WithCancel(context.Background())
+	gi := &GadgetInstance{
+		id:              instance.Id,
+		name:            instance.Name,
+		mgr:             p,
+		request:         instance.GadgetInfo,
+		eventBuffer:     make([]*bufferedEvent, 1024),
+		eventBufferOffs: 0,
+		cancel:          cancel,
+		clients:         map[*GadgetInstanceClient]struct{}{},
+	}
+	p.mu.Lock()
+	p.gadgetInstances[gi.id] = gi
+	// Adopt all clients in the waiting room
+	if p.asyncGadgetRunCreation {
+		p.waitingRoom.Range(func(key, value any) bool {
+			if value.(string) == gi.id {
+				log.Debugf("adopting client")
+				gi.AddClient(key.(api.GadgetManager_RunGadgetServer))
+			}
+			p.waitingRoom.Delete(key)
+			return true
+		})
+	}
+	p.mu.Unlock()
+	go func() {
+		defer func() {
+			cancel()
+			p.gadgetDone <- true
+		}()
+		err := gi.RunGadget(ctx, p.runtime, logger.DefaultLogger(), gi.request)
+		if err != nil {
+			log.Errorf("running gadget: %v", err)
+			gi.mu.Lock()
+			gi.state = stateError
+			gi.error = err
+			gi.mu.Unlock()
+		}
+	}()
+}
+
+func (p *Manager) LookupInstance(gadgetInstanceID string) *GadgetInstance {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	// 1. Match by complete ID
+	if gi, ok := p.gadgetInstances[gadgetInstanceID]; ok {
+		return gi
+	}
+
+	// 2. Match by name
+	for _, gi := range p.gadgetInstances {
+		if gi.name == gadgetInstanceID {
+			return gi
+		}
+	}
+
+	// 3. Partial match by ID or Name
+	for gid, gi := range p.gadgetInstances {
+		if strings.HasPrefix(gid, gadgetInstanceID) {
+			return gi
+		}
```

Comment:
Should we check if there are other instances that match the prefix? It's how it's done by docker.